### ISSUE TITLE
Can't run on non-GPU machine sometimes

### ISSUE THREAD

**Issue by soichih** [brain-life/app-mrtrix3-preproc#17]
This app was expected to run on machine with and without GPU by falling back to use eddy_openmp in case eddy_cuda fails. 

I've seen a few occasions where edd must run on GPU enabled resource given a set of configuration or input data, I believe.

```
               ============
               eddy_cuda9.1
               ============
               Reading images
               Performing volume-to-volume registration
               Running Register
               EDDY:::  EddyGpuUtils::InitGpu: cudeGetDevice returned an unknown error code
               EDDY:::  cuda/EddyGpuUtils.cu:::  static void EDDY::EddyGpuUtils::InitGpu(bool):  Exception thrown
               EDDY:::  cuda/EddyGpuUtils.cu:::  static std::shared_ptr<EDDY::DWIPredictionMaker> EDDY::EddyGpuUtils::LoadPredictionMaker(const EDDY::EddyCommandLineOptions&, EDDY::ScanType, const EDDY::ECScanManager&, unsigned int, float, NEWIMAGE::volume<float>&, bool):  Exception thrown
               EDDY:::  eddy.cpp:::  EDDY::ReplacementManager* EDDY::Register(const EDDY::EddyCommandLineOptions&, EDDY::ScanType, unsigned int, const std::vector<float, std::allocator<float> >&, EDDY::SecondLevelECModel, bool, EDDY::ECScanManager&, EDDY::ReplacementManager*, NEWMAT::Matrix&, NEWMAT::Matrix&):  Exception thrown
               EDDY::: Eddy failed with message EDDY:::  eddy.cpp:::  EDDY::ReplacementManager* EDDY::DoVolumeToVolumeRegistration(const EDDY::EddyCommandLineOptions&, EDDY::ECScanManager&):  Exception thrown
               
               
               ===========
               eddy_openmp
               ===========
               EddyInputError:  Slice-to-vol is not yet implemented for the CPU version.
               Terminating program
```

`reslice` is set to 0.8 instead of default empty, and I believe that makes this App to require GPU. Am I correct to assume that? @bacaron @bcmcpher

**Comment by bcmcpher** [comment]
Yeah, anything that passes the slice timing information for the intravolume motion within eddy requires the CUDA version of eddy. They apparently can't implement that correction without CUDA b/c it would always take too long (days) to run.

It would be nice if they just ignored the arguments if the hardware isn't available and conveniently fail over (like they do with openmp), but apparently the script doesn't do that.

What reslice argument? The one for the app or is there a specific one for topup/eddy? The reslice for the general app doesn't call CUDA (or even FSL). That will set the final DWI resolution to a 0.8 mm isotropic voxel size several steps after eddy/topup runs and isn't used for any step before.

**Comment by soichih** [comment]
Sorry I meant `eddy_mporder` option - not reslice. 

And the description says this ..
> Slice-to-vol motion correction is computationally very expensive so it is only implemented for the CUDA version.

Unless we run mrtrix3-preproc only on GPU enabled machine, user doesn't get to decide if their job will run on CPU or GPU so this is a problem.

I believe it is too expensive to always run this App on GPU.. but maybe we don't have a choice? Should we create 2 different versions of this App?

**Comment by bcmcpher** [comment]
`eddy_mporder` being set is what forces the intra-volume motion correction to require the GPU version of CUDA. `--slspec` is also required (expected?) and is needed for the GPU version to succeed. So if you can filter by those input arguments being present (somehow) that will guarantee whether the job requires a GPU or not. The app can be split w/ modified argument to ensure it's always one version or the other if that's the easiest way to solve this. If people don't want to look up / provide the slice info the app can stay CPU bound for them. 

The problem is there's never really going to be a case were someone should choose to not run the slice-to-vol motion correction. Every DWI sequence can have this step run in preprocessing if the slice order is known, which is pretty basic metadata. MRTrix3 documentation for this step is pretty new and the slice order is not easily imported / stored with the DWI data or able to be passed to the app directly, which makes it hard to add as a general input: 
https://mrtrix.readthedocs.io/en/latest/dwi_preprocessing/dwifslpreproc.html#using-eddy-s-slice-to-volume-motion-correction-capability

Multiple apps is probably a short-term fix, figuring out how to import the slice order for general use may be a longer term goal.

**Comment by soichih** [comment]
Thanks. This was interesting

> If DICOM conversion & all subsequent processing is performed solely using MRtrix3 commands, and Header key-value pairs are preserved, then where possible, MRtrix3 will store fields “SliceEncodingDirection” and “SliceTiming” based on DICOM information (note that the naming of these fields is consistent with the BIDS specification). dwifslpreproc will then use these fields to internally generate the “slspec” file required by eddy without user intervention; as long as the user does not provide the --slspec= option within -eddy_options.

What this means is that, we should not set slspec parameter if the SliceEncodingDirection*SliceTiming headers are present. (Do they mean nifti headers?)

For the timing, I think we should add checks the following condition at the beginning of the App?

1) If SliceEncodingDirection / SliceTiming is set, make sure slspec is not set.
2) If SliceEncodingDirection / SliceTiming is set, or mporder is set, then make sure that the App is currently running on GPU enabled machine (just check for CUDA91 ENV parameter).

We should ask @jdtournier if CPU version of intra-volume correction can be implemented anytime soon..

**Comment by jdtournier** [comment]
> We should ask @jdtournier if CPU version of intra-volume correction can be implemented anytime soon..

OK, I'm not sure what you have in mind here, but just to be clear: 
- we (the [MRtrix3](https://github.com/MRtrix3) team) only provide a wrapper script (`dwifslpreproc`) around the `eddy` command, which is part of [FSL](https://fsl.fmrib.ox.ac.uk/fsl/fslwiki). If you want a CPU version of that, the person you'll want to talk to is [Jesper Anderson](https://www.win.ox.ac.uk/people/jesper-andersson) - though I'm not sure he's on GitHub. And more to the point, I doubt they would invest the time in implementing this if it does genuinely takes days on the CPU... 
- for the fine detail on how all this works in terms of header entries and integration of our scripts with these FSL tools, the person to talk to is @Lestropie. But in general: yes, it's harder than it should be to extract this information from DICOM headers, and also to make sure that information stays in the headers during the processing pipeline.
- If what you're referring to here is the more [recent work of Daan Christiaens](https://www.sciencedirect.com/science/article/pii/S1053811920309228), then I'll let @dchristiaens answer that...

**Comment by bcmcpher** [comment]
Thanks for clarifying @jdtournier 

That's why I think this may need a combined short/long term. Those fields are DICOM, there's nothing in the nifti file header that can store them. The closest thing in nifti is the field for the slice order, but it can only be a (single digit?) numeric code. At one point there convention was to use single digit codes for ascending/descending/interleaved ascending/interleaved descending, but that went out when multi-banded sequences became a thing. And that doesn't deal with SliceEncodingDirection at all. 

I'd have to add this info to the initial file conversion within the app (its not currently part of the import and needs user inputs for this data) for it to know how to run, and at that point the resource has already been selected. So it can't wait that long to direct where it runs, right? The imported .mif files aren't stored long term anywhere. So I think separate apps is the easiest fix for now. Just require the arguments for the GPU version. 

The only convention I'm aware of that stores this data now is a .json sidecar in BIDS, and that stores the raw timing of the slice order (it needs to be indexed/sorted to be used). I know some files have extra fields in their metadata on brainlife, but I don't know how I'd query them through an app to check if fields like these are present. It's worth figuring out something for this and new features coming (SHARD).

For reference, FSL wants this layout for the slice order:
https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/eddy/UsersGuide#A--slspec

But I've also seen a single array notation that stores the order each slice was acquired. This keeps the dimensions consistent for single- / multi-band data, but I'm not sure who/what uses it.

**Comment by Lestropie** [comment]
I've consistently been in two minds about how to wrap what functionality where. 

*MRtrix3*'s `dwifslpreproc` will simply feed `eddy` with whatever command-line options it's instructed to pass down to it via `-eddy_options`, and so absolves itself of responsibility if a particular version of `eddy` doesn't support those options. 

In another project I [directly query](https://github.com/BIDS-Apps/MRtrix3_connectome/blob/0.5.0/mrtrix3_connectome.py#L80-L111) the help page of the installed & operational version of `eddy`, and decide there what options to instruct `dwifslpreproc` to pass down. What's annoying here though is that the OpenMP version *offers* the `--mporder` option, it merely exits if such is specified; if it were instead *omitted* from the help page, that would allow wrappers to make better decisions about what to do without making assumptions.

Slice timing can be stored in a sidecar JSON, or in a `.mif` header: they essentially serve the same purpose, though the latter is a simple string-string key-value mapping which makes *my* life a headache (MRtrix3/mrtrix3#1738 and linked issues). 

If using `dwifslpreproc`, slice timing is internally converted into the "slspec" format required by `eddy` via [this glorious one-liner](https://github.com/MRtrix3/mrtrix3/blob/3.0.2/bin/dwifslpreproc#L871).
